File: medium/typescript/36-valid-sudoku.ts
function isValidSudoku(board: string[][]): boolean {
  const cols = new Map<number, Set<string>>();
  const rows = new Map<number, Set<string>>();
  const squares = new Map<string, Set<string>>();

  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      const val = board[r][c];
      if (val === ".") {
        continue;
      }

      const squareKey = `${Math.floor(r / 3)}-${Math.floor(c / 3)}`;

      if (!rows.has(r)) rows.set(r, new Set());
      if (!cols.has(c)) cols.set(c, new Set());
      if (!squares.has(squareKey)) squares.set(squareKey, new Set());

      if (
        rows.get(r)!.has(val) ||
        cols.get(c)!.has(val) ||
        squares.get(squareKey)!.has(val)
      ) {
        return false;
      }

      rows.get(r)!.add(val);
      cols.get(c)!.add(val);
      squares.get(squareKey)!.add(val);
    }
  }

  return true;
}

File: medium/python/36-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        rows: dict[int, set[str]] = {}
        cols: dict[int, set[str]] = {}
        squares: dict[tuple, set[str]] = {}

        for r in range(9):
            for c in range(9):
                val = board[r][c]

                if val == ".":
                    continue

                if r not in rows:
                    rows[r] = set()
                if c not in cols:
                    cols[c] = set()

                square_key = (r // 3, c // 3)
                if square_key not in squares:
                    squares[square_key] = set()

                if val in rows[r] or val in cols[c] or val in squares[square_key]:
                    return False

                rows[r].add(val)
                cols[c].add(val)
                squares[square_key].add(val)

        return True
